enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,473 @@
+
+- The Operator and all platform cloud controllers will be provisioned in `openshift-cloud-controller-manager` namespace.
+
+#### Resource changes
+
+#### Kubelet
+
+OpenShift manages `kubelet` configuration with `machine-config-operator`. The actual `cloud-config` arguments are set externally, based on the `Infrastructure` [resource values](https://github.com/openshift/machine-config-operator/blob/2a51e49b0835bbd3ac60baa685299e86777b064f/pkg/controller/template/render.go#L310-L357), where the support for `external` cloud providers will be added. For each platform with `external` configuration, the `cloud-config` flag will not be set.
+
+Kubelet is tightly coupled with `CCMs`, due to its [usage](https://github.com/kubernetes/kubernetes/blame/323f34858de18b862d43c40b2cced65ad8e24052/pkg/kubelet/cloudresource/cloud_request_manager.go#L98-L102) of cloud-provider interfaces, such as `Instances` and `NodeAdresses`. Whenever there is a need to bootstrap any `Node`, `kubelet` will collect missing information about the instance IP addresses, type and zone. This information could later be used by `kube-scheduler`.
+
+During this procedure all new `Nodes` in the cluster are [tainted](https://github.com/kubernetes/kubernetes/blob/c53ce48d48372c30053a9b67c6d1dee237b5cd69/pkg/kubelet/kubelet_node_status.go#L307-L311) with `node.cloudprovider.kubernetes.io/uninitialized: true`. This taint is removed after the [initialization](https://github.com/kubernetes/kubernetes/blob/46d522d0c8e94edb6d29606a28785e35259badd7/staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go#L374-L376) by `CCM`.
+
+#### Kube Controller Manager and Kube API Server
+
+Configuration of these components is managed by `cluster-kube-controller-manager-operator` and `cluster-kube-apiserver-operator` respectively. Both operators use [`configobserver`](https://github.com/openshift/library-go/blob/16d6830d0b80dc2a3315207116d009ed2dd4cebf/pkg/operator/configobserver/cloudprovider/observe_cloudprovider.go#L140) module from `library-go` where we set what cloud provider should be used for the given platform. When a platform is being switched to `CCM`, we [must not](https://kubernetes.io/docs/tasks/administer-cluster/running-cloud-controller/#running-cloud-controller-manager) set `cloudProvider` value there and leave it empty.
+
+Next step is to bump `library-go` for the operators.
+
+##### Pre-release / Development
+
+In order to achieve desired cluster configuration for cloud components during development, we need to add an annotation on the `Infrastructure` [resource](https://github.com/openshift/api/blob/c2f7aea5d89ee87d7510b50bdb08dde9028a53ef/config/v1/types_infrastructure.go#L11).
+
+Proposed annotation: `infrastructure.openshift.io/external-cloud-provider`.
```

Comment:
This is not going to be accepted by @openshift/api-approvers. We already sufficient information which platform we are running on. Based on that, we should gradually set appropriate cloud-config values for all components and enable the necessary elements of ccm. I imagine this will take several releases, but at any given release we will document which external providers are supported in-tree and which out-of-tree.

Reply:
I doubt that there will be half-implemented solution, as proposed through annotation. For example, for openstack we'll move all components (kas, kcm, kubelet) to not to set cloud-config flag.

Reply:
I think the intention of this is that it shouldn't make it into a release and that it shouldn't be publicly available or documented.

During the development cycle, we will need to update various components to work with out of tree providers, if we implement these one by one and switch (eg update KCM to remove the flag), then this will break clusters and will not pass E2E.

The idea is that once all of the components have implemented their changes, we will be able to test all of the components together by adding the annotation, somehow add it to new clusters as part of the install process, then update all of the components to remove the switch, and then remove the annotation. 

I believe this is what this section is trying to explain.

Do you know of a better way to co-ordinate the changes across multiple components like this?

Reply:
Was discussing this with @Fedosin and @Danil-Grigorev again and now that the plans have changed slightly (OSP TP in 4.7) it looks like we are going to need some way for a user to opt-in to having an external cloud provider for 4.7. I imagine however we implement this, it can replace the usage of the annotation as we have described it in this proposal.

I'm not sure what the standard practice for this is within OpenShift, is there some resource a user can create or some place where they can opt in to tech previews such as this?

I think the process we want to have is:
- User indicates via some resource that they want to enable out of tree
- KCMO, KASO, MCO observe this change and reconfigure the respective components to enable out of tree
- CCCMO observes this change and deploys the relevant CCM Deployment

Since there will be no CSI migration enabled for the TP for OSP, this will have to come with a proviso that all existing PVs will be lost, not sure how we communicate that.

As Out of tree moves to GA in future releases, the above components will just look at the `platform` value to determine which configuration they should be using on a given platform and this resource/field will no longer be required.

I'm imagining some logic something like:

```
if (featureFlagIsSet && setOfPlatformsInDevOrTP.Has(infrastructure.Platform)) || setOfPlatformsInGA.Has(infrastructure.Platform) {
  // Deploy out of tree configuration for given provider
} else {
  // Deploy in tree configuration for provider
}
```

The question is, where does the feature flag live? How is that implemented?

Reply:
o/api/config/v1 FeatureGate